### PR TITLE
Adding examples for creating custom property packages

### PR DESCRIPTION
**Pull request by andrewlee94**
## Proposed changes:
- Add introductory examples for creating custom thermophysical and reaction property packages.

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.txt file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

**Comment by andrewlee94** [comment]
@TimBartholomew @jghouse88 The first draft for the physical properties example is now finished, with all the text added. Comments are welcome.

**Review by TimBartholomew — COMMENTED** [review]
@andrewlee94 thank you for compiling all of this, as you know these tutorials were really needed.

I have three types of comments: 1) general, 2) typos, and 3) suggested rewordings. Here I'm just going to include the general comments. I'll discuss with you on how best to share the typos and rewording suggestions, I would really like something like the tracked changes feature in Microsoft Word, but I don't think that's going to be possible for a Jupyter notebook.

1.  I like the structure and order of the tutorial (i.e. headings and subheadings). However, I think the content could be more clear and succinct through three ways. 
    - I would shift from a discussion with the reader to just telling the reader (i.e. removing side comments). There are several sections that walk through the thought process or consider multiple ways to do something, but then end with the recommended approach. For a tutorial like this, which is already pretty long, I would just tell them what it is or how to do it and refer them to the other documentation you made on custom property models.
    - I would make the writing more consistently active first person. Currently, the tutorial includes a significant mix of second and third person (i.e. you and they). I think writing with just the general first person (i.e. inclusive we) will be more clear. The tutorial also switches from section to section on whether active or passive voice is the most dominant. I think active voice should be the consistent dominant voice with passive voice used only when it helps the flow of the paragraph. If you like, I could take a shot at this for the whole document with my suggested rewording comments.
    - I would cut out unnecessary clauses. For example, in the note under the "What is a Property?" section, it starts with "Note: It is important to emphasize that extensive flow information is considered" it could just be replaced with "Note: Extensive flow information is considered". These types of things will be noted in my suggested rewording comments.
2. There are some inconsistencies with formatting. For example, should it be State Block or StateBlock or StateBlock? I see same thing with Physical Parameter Block. I think things should only be italicized when we are specifically pointing to a class, method, or variable, rather than when we are generally discussing that thing (so it could still be consistent to use State Block sometimes and StateBlock other times depending on the context, but I don’t think the current tutorial has done this). An example of something I think should be italicize is the flowsheet and unit model under the “How are Properties Used?” section. Along these same lines I don’t think we should italicize general python terms, for example, in the first sentence under the Importing Libraries section “classes” is italicized. Another example is “dict” is italicized throughout, but I would just write out dictionary (i.e. “units_metadata dict” should be “units_metadata dictionary”).
3. In the “Define Supported Properties” section, the example given in the text for using None in the properties_metadata includes specific enthalpy, but in cell 3, specific enthalpy is the only one that has a method argument. I would use a different example of something that would be useful to not be built on demand. 
4. In the “Declaring the Physical Parameter Block” section, I like how the 4 steps are listed at the beginning, but then I got lost a little in the following paragraphs. I think it would be more clear if the first sentence of each paragraph repeated or paraphrased the step and then the following content was more streamlined.
5. In the “State Block” section, I think it needs to start with a summary/description of the state block. Currently it quickly goes into details like two classes and the indexing, but people should be reminded what a state block is and its purpose before that.
6. The state block content before step 5 is fairly hard to follow. I think you could take two different approaches: 1) describe the differences of IndexedComponent and ComponentData objects with a figure and significantly revise the current content, 2) skip over the differences and just state what each class is for (i.e. methods applied to all StateBlocks at once, methods applied to an individual StateBlock) and refer the user to your more detailed documentation on custom property models.
7. I’m not sure that quoting state variables or state is necessary.
8. The terms in cell 11 are in a different order than the presented equation, to make it consistent I would reorder the terms in cell 11. 
9. Why are we scaling the mole fraction constraint? It’s value is 1, which is already well scaled, right? Also I thought this type of formulation is discouraged and that the mole fractions should instead be calculated from the component molar flowrates.
10. I’m unsure we need all the initialization content, and I think it might distract from the tutorial. I think it is ok to describe the overall approach and what each part does, but it currently includes a lot of discussion and alternatives. I think that discussion should move to a separate initialization document. Also depending on if we reformulate the package to not include the mole fraction constraint, then the initialization would be completely generic.

File: src/Examples/Advanced/CustomProperties/custom_physical_property_packages.ipynb [hunk 1]
```diff
@@ -0,0 +1,1329 @@
...
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "m = ConcreteModel()\n",
+    "\n",
+    "m.fs = FlowsheetBlock(default={\"dynamic\": False})\n",
+    "\n",
+    "m.fs.thermo_props = HDAParameterBlock()\n",
+    "\n",
+    "m.fs.state = m.fs.thermo_props.build_state_block(\n",
```
Comment (adam-a-a) [inline src/Examples/Advanced/CustomProperties/custom_physical_property_packages.ipynb:973]: Here, in code block 17, I hit an error: 
`PropertyNotSupportedError: fs.state[0.0] package property metadata method does not contain a method for component_list. Please select a package which supports the necessary properties for your process.`

Comment (adam-a-a) [reply src/Examples/Advanced/CustomProperties/custom_physical_property_packages.ipynb:973]: The traceback is more extensive but didn't want to leave such a verbose comment.

Comment (andrewlee94) [reply src/Examples/Advanced/CustomProperties/custom_physical_property_packages.ipynb:973]: I don't get an issue when I run the example. Could you check your IDAES version first of all, and also send me the full traceback (Slack is fine).

Comment (adam-a-a) [reply src/Examples/Advanced/CustomProperties/custom_physical_property_packages.ipynb:973]: Resolved this issue by using a later version of IDAES, but now the last cell gives an error; it appears that the RPP property package no longer exists (instead I see RPP3, RPP4, and RPP5 in  `idaes/generic_models/properties/core/pure` 

I'd say that incorporating `doctest` would be useful, except I'm not sure how that would work for a jupyter notebook; I do know how to work doctest with .rst files though.

Comment (andrewlee94) [reply src/Examples/Advanced/CustomProperties/custom_physical_property_packages.ipynb:973]: We have other ways of testing Jupyter notebooks, and we will include some test cells in the final version.

Regarding the `RPP` module, I probably updated that on my local branch and it is a small change.

File: src/Examples/Advanced/CustomProperties/custom_physical_property_packages.ipynb [hunk 2]
```diff
@@ -0,0 +1,1329 @@
...
+    "\n",
+    "### The StateBlock class\n",
+    "\n",
+    "To begin, we declare a new class which inherits from the IDAES `StateBlock` base class, which we will call `_HDAStateBlock` here. Note that we need to be careful with naming here; later we will be automatically generating a class names `HDAStateBlock` which will be the actual state block that we use for our model (and incidentally the one we created a pointer to back in the Physical Parameter Block class). Thus, we add an underscore to the beginning of this class name to distinguish it from the class that will be generated later, and to indicate that this is for internal use only by the property package.\n",
+    "\n",
+    "Next, the `_HDAStateBlock` class needs to have two methods added to it; `initialize` and `release_state` which are used to run the initialization routine for the State Block. Each of these have a set of arguments which must part of the method definition (you can add more if you have a need for them, but the pre-defined set are a minimum).\n",
+    "\n",
+    "#### The `initialize` method\n",
```
Comment (adam-a-a) [inline src/Examples/Advanced/CustomProperties/custom_physical_property_packages.ipynb:799]: The `initialize` method just needs a quick update to (1) match the new usage of `get_solver()` and (2) set default arguments to `None` rather than `{}`

File: src/Examples/Advanced/CustomProperties/custom_physical_property_packages.ipynb [hunk 3]
```diff
@@ -0,0 +1,1329 @@
...
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Concluding Remarks\n",
+    "\n",
+    "The above example has hopefully introduced you to the basic requirements for creating your own custom physical property packages. However, it is probably clear that it requires a significant amount of effort to write your own property packages, thus users are encouraged to look into the IDAES Generic Properties Framework if they are not already familiar with this.\n",
+    "\n",
+    "The  IDAES Generic Properties Framework is designed to automatically generate user-defined property packages for common applications based on a single configuration file. User provide a list of phase and components of interest, and select fro ma library of common forms for properties of interest, and the Generic Properties Framework then does the hard work of assembling the necessary code to construct the desired model.\n",
+    "\n",
+    "To demonstrate the application of the Generic Properties Framework, an equivalent property package to the one we wrote above can be generated using the following configuration dict."
```
Comment (adam-a-a) [inline src/Examples/Advanced/CustomProperties/custom_physical_property_packages.ipynb:1210]: Is the application of the Generic Properties Framework fully presented in the final code cell? I see that a config dictionary is setup. Is there anything else that was meant to be shown after that in the code?

**Comment by adam-a-a** [comment]
Thus far, I went through the Jupyter notebook for custom physical property packages and pushed my suggested changes to @andrewlee94 's branch; I took care of all typos (unless I missed any), made other very subtle revisions, added hyperlinks for references to other IDAES documentation, and tweaked one module name in a code cell that needed to be updated. 

In response to @TimBartholomew 's comments above, my general thought is that we should aim to release this documentation ASAP so that it could be put to use sooner rather than later. While the example that I went through is quite verbose and could probably be repackaged in the future (e.g., save this version as "detailed explanation/background", strip out everything but the section headings and the necessary code cells and make that a "tutorial," etc.) , it is also extremely informative and helpful. With regard to typos and rewording, I think I took care of everything that really stood out. Any other points, IMHO, can be addressed in a future PR. 

I still need to go through the other notebook on custom reaction property packages and do the same that I did for the one on custom physical property packages, but once that is done, I would move towards approving and merging this PR. I am also going to tag @bpaul4 here because, like me, he is one of the newer members of our team, and I think he could benefit from scanning through these notebooks while giving his take on things.

File: src/Examples/Advanced/CustomProperties/custom_physical_property_packages.ipynb (hunk 3 above)
Comment (andrewlee94) [reply src/Examples/Advanced/CustomProperties/custom_physical_property_packages.ipynb:1210]: I was not intending to show anything else - just how to create the config dict. Docs for the generic properties will be else where.

**Review by bpaul4 — COMMENTED** [review]
As a new user, I appreciated the level of detail in the Physical Property Package tutorial and look forward to viewing the Reaction Property Package tutorial when it is complete. In this example data was entered manually for demonstration purposes, and I believe it may be useful to demonstrate importing external VLE data (read_csv, read_excel, or using ParamEst) even if just a quick example showing that it can be done for custom packages. I'm also about to do a pull request on @andrewlee94 's branch for a few typos I came across.

**Comment by andrewlee94** [comment]
@bpaul4 I am not quite sure what you mean by importing external data here, but I think  things like that would be topics for separate examples.

**Comment by andrewlee94** [comment]
@TimBartholomew @adam-a-a @bpaul4 I have filled in the text to go with the reaction property example, and hopefully address a lot of the critical issues. It would be appreciated if you could take the time to look over the changes.

File: src/Examples/Advanced/CustomProperties/custom_reaction_property_packages_testing.ipynb [hunk 4]
```diff
@@ -0,0 +1,720 @@
...
+    "\n",
+    "Next, we create a `dict` object for the stoichiometric coefficients. This `dict` needs to provide coefficients for all combinations of reaction, phase and component present in the system, even for those components which do not take part in a reaction. This is required as `ControlVolumes` create generation terms for all reaction-phase-component combinations, and need a stoichiometric coefficient for each of these. In this `dict`, the keys need to have the form of a tuple with three parts:\n",
+    "\n",
+    "1. the reaction name,\n",
+    "2. the phase name, and\n",
+    "3. the component name,\n",
+    "\n",
+    "whilst the value is the stoichiometric coefficient for that key combination. See the example in the cell belo; in this example we have 1 reaction (`R1`), 1 phase (`Vap`, as defined in the thermophysical property package) and 5 components (`benzene`, `toluene`, `hydrogen`, `methane` and `diphenyl`), thus the resulting dict has `1x1x5` entries."
```
Comment (adam-a-a) [inline src/Examples/Advanced/CustomProperties/custom_reaction_property_packages_testing.ipynb:180]:
```suggestion
    "whilst the value is the stoichiometric coefficient for that key combination. See the example in the cell below; in this example we have 1 reaction (`R1`), 1 phase (`Vap`, as defined in the thermophysical property package) and 5 components (`benzene`, `toluene`, `hydrogen`, `methane` and `diphenyl`), thus the resulting dict has `1x1x5` entries."
```

File: src/Examples/Advanced/CustomProperties/custom_reaction_property_packages_testing.ipynb [hunk 5]
```diff
@@ -0,0 +1,720 @@
...
+    "\n",
+    "Next, we need to declare the `build` method that will be sued to construct our Reaction Parameter Block. The first step in any `build` method is to call `super().build()`, which will trigger the `build` method of the base class that the current class inherits from – this is important since this is how we automate construction of any underlying components required by the modeling framework and ensure that everything integrates smoothly. Next, a `ReactionParameterBlock` needs to contain a pointer to the related `ReactionBlock` (which we will look at next) – this is used to allow us to build instances of the `ReactionBlock` by only knowing the `ReactionParameterBlock` we wish to use. To do this, we create an attribute named `_reaction_block_class` attached to our class with a pointer to the `ReactionBlock` class; in this case `self._reaction_block_class = HDAReactionBlock`, where `HDAReactionBlock` is the name of the yet to be declared `ReactionBlock`. Finally, the `build` method needs to construct the actual parameters required for the property package, which we do here by calling the sub-methods written previously.\n",
```
Comment (bpaul4) [inline src/Examples/Advanced/CustomProperties/custom_reaction_property_packages_testing.ipynb:268]: 'sued' should be 'used'

**Review by bpaul4 — COMMENTED** [review]
Again, I appreciate the level of detail in the two custom property package examples and especially the explanation of the various definitions and calls in the package. Besides the typo @adam-a-a found on line 180, I found another on line 268 ('sued' instead of 'used').

File: src/Examples/Advanced/CustomProperties/custom_reaction_property_packages_testing.ipynb [hunk 6]
```diff
@@ -0,0 +1,720 @@
...
+    "$$\n",
+    "\\text{Toluene} + \\text{Hydrogen} \\rightarrow \\text{Benzene} + \\text{Methane}\n",
+    "$$\n",
+    "$$\n",
+    "2 \\text{Benzene} \\rightleftharpoons \\text{Hydrogen} + \\text{Diphenyl}\n",
+    "$$\n",
+    "\n",
+    "## A Note on this Tutorial\n",
+    "\n",
+    "The `build` methods in the reaction property package classes are generally written as a single, long method. However, to break the code in the manageable pieces for discussion, in this tutorial we will create a number of smaller sub-methods that will then be called as part of the `build` method. This is done entirely for presentation purposes, and model developers should not feel compelled to write their models this way.\n",
```
Comment (adam-a-a) [inline src/Examples/Advanced/CustomProperties/custom_reaction_property_packages_testing.ipynb:68]:
```suggestion
    "The `build` methods in the reaction property package classes are generally written as a single, long method. However, to break the code into manageable pieces for discussion, in this tutorial we will create a number of smaller sub-methods that will then be called as part of the `build` method. This is done entirely for presentation purposes, and model developers should not feel compelled to write their models this way.\n",
```

File: src/Examples/Advanced/CustomProperties/custom_reaction_property_packages_testing.ipynb [hunk 7]
```diff
@@ -0,0 +1,720 @@
...
+    "\n",
+    "The final step in creating the `ReactionParameterBlock` class is to declare a `classmethod` named `define_metadata` which takes two arguments; a class (`cls`) and an instance of that class (`obj`). This method in turn needs to call to pre-defined methods (inherited from the underlying base classes):\n",
```
Comment (adam-a-a) [inline src/Examples/Advanced/CustomProperties/custom_reaction_property_packages_testing.ipynb:270]:
```suggestion
    "The final step in creating the `ReactionParameterBlock` class is to declare a `classmethod` named `define_metadata` which takes two arguments; a class (`cls`) and an instance of that class (`obj`). This method in turn needs to call two pre-defined methods (inherited from the underlying base classes):\n",
```

File: src/Examples/Advanced/CustomProperties/custom_reaction_property_packages_testing.ipynb [hunk 8]
```diff
@@ -0,0 +1,720 @@
...
+    "\n",
+    "## State Variables\n",
+    "\n",
+    "Like thermophysical property calculations, reaction properties also depend on the material state variables such as temperature and pressure. However, the state variables are declared as a part of the State Block, and it does not make sense to duplicate them here. Due to this, Reaction Blocks are always associated with a State Block representing the material at the same point in space and time, and the Reaction Block contains a pointer to the equivalent State Block. This allows the Reaction Block to access the state variables, and any other thermophysical property the State Block supports, in order to perform reaction property calculations. The State Block can be accessed using the `state_ref` property of the Reaction Block.\n",
+    "\n",
+    "\n",
+    "## Step 1. Define Property Variables\n",
+    "\n",
+    "The first thing we need to do when creating our Reaction Block is create Pyomo component to represent the properties of interest. In this example, we have three properties we need to define:\n",
```
Comment (adam-a-a) [inline src/Examples/Advanced/CustomProperties/custom_reaction_property_packages_testing.ipynb:323]:
```suggestion
    "The first thing we need to do when creating our Reaction Block is create Pyomo components to represent the properties of interest. In this example, we have three properties we need to define:\n",
```

File: src/Examples/Advanced/CustomProperties/custom_reaction_property_packages_testing.ipynb [hunk 9]
```diff
@@ -0,0 +1,720 @@
...
+    "\n",
+    "1. the rate constant for the rate-based reaction: `k_rxn`,\n",
+    "2. a variable for the rate of reaction at the current state, `rate_reaction`, and\n",
+    "3. the equilibrium constant for the equilibrium-based reaction.\n",
```
Comment (adam-a-a) [inline src/Examples/Advanced/CustomProperties/custom_reaction_property_packages_testing.ipynb:327]:
```suggestion
    "3. the equilibrium constant, `k_eq`, for the equilibrium-based reaction.\n",
```

File: src/Examples/Advanced/CustomProperties/custom_reaction_property_packages_testing.ipynb [hunk 10]
```diff
@@ -0,0 +1,720 @@
...
+    "                      units=pyunits.Pa)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Step 2. Define Constraints for the Rate-Based Reactions\n",
+    "\n",
+    "Next, we need to define the `Constraints` which describe the rate-based reaction. First, we use the Arrhenius equation to calculate the rate constant, $k_{rxn} = A \\times e^{E_a/(RT)}$. For this calculation, $A$ and $E_a$ come from the associated Reaction Parameter Block (`self.params`), $T$ comes from the associated State Block (`self.state_ref.temperature`) and the gas constant $R$ can be found in the IDAES `Constants` class.\n",
```
Comment (adam-a-a) [inline src/Examples/Advanced/CustomProperties/custom_reaction_property_packages_testing.ipynb:359]:
```suggestion
    "Next, we need to define the `Constraints` which describe the rate-based reaction. First, we use the Arrhenius equation to calculate the rate constant, $k_{rxn} = A \\times e^{-E_a/(RT)}$. For this calculation, $A$ and $E_a$ come from the associated Reaction Parameter Block (`self.params`), $T$ comes from the associated State Block (`self.state_ref.temperature`) and the gas constant $R$ can be found in the IDAES `Constants` class.\n",
```

File: src/Examples/Advanced/CustomProperties/custom_reaction_property_packages_testing.ipynb [hunk 11]
```diff
@@ -0,0 +1,720 @@
...
+    "assert value(m.fs.reactor.outlet.flow_mol[0]) == pytest.approx(100, abs=1e-3)\n",
+    "assert value(m.fs.reactor.outlet.temperature[0]) == pytest.approx(790.212, abs=1e-3)\n",
+    "assert value(m.fs.reactor.outlet.mole_frac_comp[0, \"benzene\"]) == pytest.approx(0.159626, abs=1e-6)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Finally, as a quick check of model consistency, let us assert that the units of measurement in our model are consistent (the units of the rate constant and pre-exponential factor at rather complex)."
```
Comment (adam-a-a) [inline src/Examples/Advanced/CustomProperties/custom_reaction_property_packages_testing.ipynb:674]:
```suggestion
    "Finally, as a quick check of model consistency, let us assert that the units of measurement in our model are consistent (the units of the rate constant and pre-exponential factor are rather complex)."
```

File: src/Examples/Advanced/CustomProperties/thermophysical_property_example.py [hunk 12]
```diff
@@ -0,0 +1,380 @@
...
+# Please see the files COPYRIGHT.txt and LICENSE.txt for full copyright and
+# license information, respectively. Both files are also available online
+# at the URL "https://github.com/IDAES/idaes-pse".
+##############################################################################
+"""
+Example of a thermophysical property package for the hydrodealkylation of
+toluene to form benzene, along with the side reaction to form diphenyl.
+
+This package assumes a single, ideal vapor phase and contains calculations
+for molar denisty, heat capacity and specific enthalpy.
```
Comment (adam-a-a) [inline src/Examples/Advanced/CustomProperties/thermophysical_property_example.py:18]:
```suggestion
for molar density, heat capacity and specific enthalpy.
```

File: src/Examples/Advanced/CustomProperties/thermophysical_property_example.py [hunk 13]
```diff
@@ -0,0 +1,380 @@
...
+            if hold_state is True:
+                return flags
+            else:
+                blk.release_state(flags)
+
+        init_log.info("Initialization Complete")
+
+    def release_state(blk, flags, outlvl=idaeslog.NOTSET):
+        '''
+        Method to relase state variables fixed during initialization.
```
Comment (adam-a-a) [inline src/Examples/Advanced/CustomProperties/thermophysical_property_example.py:276]:
```suggestion
        Method to release state variables fixed during initialization.
```

File: src/Examples/Advanced/CustomProperties/thermophysical_property_example.py [hunk 14]
```diff
@@ -0,0 +1,380 @@
...
+
+        init_log.info_high("State Released.")
+
+
+@declare_process_block_class("HDAStateBlock",
+                             block_class=_HDAStateBlock)
+class HDAStateBlockData(StateBlockData):
+    """
+    Example property package for an ideal gas containing benzene, toluene
+    hydrogen, mehtane and diphenyl.
```
Comment (adam-a-a) [inline src/Examples/Advanced/CustomProperties/thermophysical_property_example.py:303]:
```suggestion
    hydrogen, methane and diphenyl.
```

File: src/Examples/Advanced/CustomProperties/thermophysical_property_example.py [hunk 15]
```diff
@@ -0,0 +1,380 @@
...
+        self.mw_comp = Reference(self.params.mw_comp)
+
+        if self.config.defined_state is False:
+            self.mole_fraction_constraint = Constraint(
+                expr=1e3 == sum(1e3*self.mole_frac_comp[j]
+                                for j in self.component_list))
+
+        self.dens_mol = Var(initialize=1,
+                            units=pyunits.mol/pyunits.m**3,
+                            doc="Mixture denisty")
```
Comment (adam-a-a) [inline src/Examples/Advanced/CustomProperties/thermophysical_property_example.py:339]:
```suggestion
                            doc="Mixture density")
```

**Review by adam-a-a — CHANGES_REQUESTED** [review]
This PR will be very valuable to users once merged. I only point out subtle typos that need correction. Once those are fixed, I think this PR will be ready to merge.

**Comment by andrewlee94** [comment]
@adam-a-a @bpaul4 I believe I fixed all the typos you found. so hopefully this is ready to approve and merge.

**Review by adam-a-a — APPROVED** [review]
The time has come!

**Review by bpaul4 — APPROVED** [review]
Great work on the notebooks, and thank you @andrewlee94 and @adam-a-a for including me in the review process.
